ait/core/server/stream.py

```diff
@@ -0,0 +1,90 @@
+from client import ZMQInputClient, PortInputClient, PortOutputClient
+
+
+class Stream(object):
+    """
+    This is the base Stream class that all streams will inherit from.
+    It calls its handlers to execute on all input messages sequentially,
+    and validates the handler workflow if handler input and output
+    types were specified.
+    """
+
+    def __init__(self, name, input_, handlers, zmq_args={}, **kwargs):
+        self.name = name
+        self.input_ = input_
+        self.handlers = handlers
+
+        if not self.valid_workflow():
+            raise ValueError('Sequential workflow inputs and outputs ' +
+                             'are not compatible. Workflow is invalid.')
+
+        # This calls __init__ on subclass of ZMQClient
+        if 'output' in kwargs:
+            super(Stream, self).__init__(input_=self.input_,
+                                         output=kwargs['output'],
+                                         **zmq_args)
+        else:
+            super(Stream, self).__init__(input_=self.input_,
+                                         **zmq_args)
+
+    def __repr__(self):
+        return '<%s name=%s>' % (self.type, self.name)
+
+    @property
+    def type(self):
+        return type(self)
+
+    def process(self, input_data, topic=None):
+        """
+        Invokes each handler in sequence.
+        Publishes final output data.
+        """
+        for handler in self.handlers:
+            output = handler.execute_handler(input_data)
+            input_data = output
+
+        self.publish(input_data)
+
+    def valid_workflow(self):
```

Comment:
Similarly to the `validate_input` comment, do we care about this? I'm not convinced and would like to hear opinions.

Reply:
This would be helpful to the user in setting up their handlers, so they would know from the beginning if their handlers weren't able to be executed in sequence. It is an optional functionality which only checks compatibility if output and input type of next handler are both specified.